benchmark_online.py

```diff
@@ -0,0 +1,175 @@
+import torch.cuda.profiler
+import onnxruntime
+import numpy as np
+
+import models
+import text_processing
+
+
+def main(args):
+	use_cuda = 'cuda' in args.device
+	if use_cuda:
+		print(f'CUDA_VISIBLE_DEVICES: {os.environ.get("CUDA_VISIBLE_DEVICES")!r}')
+	print('initializing model...')
+
+	if args.onnx:
+		# todo: pass dict with provider setting when we will migrate to onnxruntime>=1.7
+		onnxruntime_session = onnxruntime.InferenceSession(args.onnx)
+		model = lambda x: onnxruntime_session.run(None, dict(x=x, xlen=[1.0] * len(x)))
+		load_batch = lambda x: x.numpy()
+		args.sample_rate = 8000
+	else:
+		assert args.checkpoint is not None and os.path.isfile(args.checkpoint)
+		checkpoint = torch.load(args.checkpoint, map_location='cpu')
+		args.model, args.lang, args.sample_rate, args.window_size, args.window_stride, args.window, args.num_input_features = (
+			checkpoint['args'].get(key) for key in
+			['model', 'lang', 'sample_rate', 'window_size', 'window_stride', 'window', 'num_input_features']
+		)
+
+		text_config = json.load(open(args.text_config))
+		text_pipelines = []
+		for pipeline_name in args.text_pipelines:
+			text_pipelines.append(text_processing.ProcessingPipeline.make(text_config, pipeline_name))
+
+		frontend = models.LogFilterBankFrontend(
+			args.num_input_features,
+			args.sample_rate,
+			args.window_size,
+			args.window_stride,
+			args.window,
+			stft_mode=args.stft_mode
+		)
+		model = getattr(models, args.model)(
+			num_input_features=args.num_input_features,
+			num_classes=[pipeline.tokenizer.vocab_size for pipeline in text_pipelines],
+			frontend=frontend,
+			dict=lambda logits, log_probs, olen, **kwargs: logits[0]
+		)
+		model.load_state_dict(checkpoint['model_state_dict'], strict=False)
+		model.to(args.device)
+		model.eval()
+		model.fuse_conv_bn_eval()
+		if use_cuda:
+			model, *_ = models.data_parallel_and_autocast(model, opt_level=args.fp16, data_parallel=False)
+		load_batch = lambda x: x.to(args.device, non_blocking=True)
+
+	batch_width = int(math.ceil(args.T * args.sample_rate / 128) * 128)
+	example_time = batch_width / args.sample_rate
+	batch = torch.rand(args.B, batch_width)
+	batch = batch.pin_memory()
+
+	print(f'batch [{args.B}, {batch_width}] | audio {args.B * example_time:.2f} sec\n')
+
+	def tictoc():
+		if use_cuda:
+			torch.cuda.synchronize()
+		return time.time()
+
+	print(f'Warming up for {args.warmup_iterations} iterations...')
+	tic_wall = tictoc()
+	if use_cuda:
+		torch.backends.cudnn.benchmark = True
+	torch.set_grad_enabled(False)  # no back-prop - no gradients
+	for i in range(args.warmup_iterations):
+		model(load_batch(batch))
+	print(f'Warmup done in {tictoc() - tic_wall:.1f} sec\n')
+
+	n_requests = int(round(args.benchmark_duration * args.rps))
+	print(f'Starting {args.benchmark_duration} second benchmark ({n_requests} requests, rps {args.rps:.1f})...')
+	schedule = np.random.random(n_requests) * args.benchmark_duration  # uniform random distribution of requests
```

Comment:
if these np.random.random, np.sort and np.diff, np.quantile are all numpy usages, please replace them with torch versions

Reply:
Done